proxy_agent/src/provision.rs

```diff
@@ -146,57 +201,99 @@ pub async fn get_provision_status_wait(
             }
         }
 
-        return provision_status;
+        // wait timedout return as 'not finished'
+        return (false, String::new());
     }
 }
 
-// Get provision status
+// Get current provision status from GPA service via http request
 // return value
 //  bool - true provision finished; false provision not finished
 //  String - provision error message, empty means provision success or provision failed.
-fn get_provision_status(provision_dir: Option<PathBuf>) -> (bool, String) {
-    let provision_dir = provision_dir.unwrap_or_else(config::get_keys_dir);
-
-    let status_file: PathBuf = provision_dir.join(STATUS_TAG_FILE_NAME);
-    if !status_file.exists() {
-        return (false, String::new());
+fn get_current_provision_status(port: u16) -> std::io::Result<ProivsionState> {
+    let provision_url = url::Url::parse(&format!(
+        "http://{}:{}{}",
+        Ipv4Addr::LOCALHOST,
+        port,
+        PROVISION_URL_PATH
+    ))
+    .map_err(|e| {
+        std::io::Error::new(
+            std::io::ErrorKind::Other,
+            format!("Failed to parse provision url with error: {}", e),
+        )
+    })?;
+    let mut req = Request::new(PROVISION_URL_PATH.to_string(), "GET".to_string());
+    req.headers
+        .add_header(constants::METADATA_HEADER.to_string(), "True ".to_string());
```

Comment:
Because I see GPA code various places either adding or checking different headers. Can you wrap into some helper functions, so they are in a central place. (Maybe hyper_client when you move over.)

Reply:
hyper_client has build_request function accepts hashmap for all the headers.